Pending ADR for new approach to monitoring and alerting (#1476)
alphagov/email-alert-api: docs/adr/adr-008-monitoring-and-alerting.md

incidents.

The approach used to monitoring the application and the wider email alert
system is two fold. One approach is to monitor Email Alert API's internal
state to see if it is [completing tasks within a particular
time frame][monitor-time-frame] or [whether there is a high quantity of work to
do][monitor-workload]. The other approach is to use an [end-to-end
check][email-alert-monitoring] which verifies that a Gmail account
has received an email within a time frame. Both of these approaches are flawed.

Monitoring for high quantities of work, and whether work was completed in a
time frame, has left Email Alert API vulnerable to producing alerts when the
application is busy, yet fully operational. These alerts are unactionable for
support engineers as the alerts resolve themselves once the work is completed
and there is little an engineer can do to speed this process up. For example,
if the application has accumulated 1,000,000 emails to send it raises a
critical alert even if it is sending these at the [fastest rate Notify
would allow][rate-limit]. These alerts, therefore, waste the time of support
engineers and create a risk that a support engineer might miss an alert that

> requires their intervention.

Reviewer: Minor: we could acknowledge that it is still necessary to monitor these things for capacity planning, but that this should not be conflated with alerts for when the system is (definitely) broken.

Since these alerts can't be relied upon to reliably suggest the
application is broken, they aren't used for any out-of-hours alerts. Therefore,
even if they did reflect a genuine problem no engineers would be contacted.
An engineer instead learns of a problem out-of-hours due to the other type
of monitoring approach: the end-to-end checks. These alert when content
published through [Travel Advice Publisher][] does not arrive in
a particular Gmail inbox within a specified time period.

Monitoring an inbox for a limited set of emails has its own distinct flaws.
If Email Alert API breaks outside of office hours then engineers would not be
notified unless travel advice content was published. This alert also monitors
beyond the boundaries of the email alert system - considering whether
GOV.UK publishing, Notify and Gmail are all working - leading to alerts

> representing problems outside the control of GOV.UK engineers (such as Gmail

Reviewer: ```suggestion
representing problems that are irrelevant to the system, or outside the control of GOV.UK engineers (such as Gmail
```

Minor: previously I didn't clock that we're implying problems with GOV.UK Publishing might be outside the control of GOV.UK engineers.

alert system of applications in order to meet this new alerting criteria.

### Monitoring Email Alert API

#### Sidekiq workers

We will configure critical alerts, that contact engineers out-of-hours, if
individual Sidekiq jobs fail multiple times for the following workers:

- [DailyDigestInitiatorWorker][]
- [DigestEmailGenerationWorker][]
- [ProcessContentChangeWorker][]
- [ProcessMessageWorker][]
- [SendEmailWorker][]
- [WeeklyDigestInitiatorWorker][]

These workers will use the database to store the number of times we have
attempted to perform an action. An alert will be raised if the quantity of
failed attempts reaches a threshold (the suggestion is 5 - based on this being
evidence of consistent failure). To ensure that these jobs actually run we
will configure the [RecoverLostJobsWorker][] to identify any of these that
need to run and attempt them, this protects from lost work problems such
as [Sidekiq crashing][sidekiq-reliability].

We expect [SendEmailWorker][] and [DigestEmailGenerationWorker][] jobs to run
quickly (durations of up to a few seconds) and know that there can be high
volumes of them to run (both can regularly reach volumes of > 100,000).

> Therefore, we are concerned about a scenario where these consistently fail

Reviewer: ```suggestion
Therefore, we are concerned about a scenario where these consistently fail,
```

(They're not failing because it could take a long...)

because it could take a long time for sufficient retries to occur
for an alert to be raised (as each retry is placed at the back of the queue).
To reflect this concern we will contact engineers if either of these workers
has not completed one successful run within a time period despite work
existing (suggested time period is 30 minutes). The [Sidekiq retry
cadence][sidekiq-retry] and [recovery window][] for these jobs will need to be
updated to ensure the jobs are always attempted within the time period, this
is to ensure we don't alert while work isn't being attempted.

We will contact engineers out-of-hours if we determine that
[DigestRuns][DigestRun] are not created within a reasonable time frame

> (suggestion is 2 hours). Creating these is a [trivial][digest-run-creation],

Reviewer: Minor: the wording here is slightly at-odds with the definition in the README, which describes the "batch" combo of the DigestRun object and all the DigestRunSubscribers. It may help to explain why we're describing the creation of DigestRun records separately - they are (trivially) created on a schedule, and we're monitoring the schedule.

engineer is unlikely to be contacted in the event of the whole Email Alert API
application becoming unavailable - unless travel advice were published. We
noted there wasn't a particularly consistent approach to this across GOV.UK
([with some apps raising an out-of-hours alert when a single instance fails
healthchecks][publishing-api-alert]) and felt it would be good to take a simple
approach to avoid surprising engineers.

We decided the most appropriate place to alert would be when [the application
load balancer has no healthy instances of the application][aws-lb-check]. This
represents that no instances of the web application are healthy and
[no requests will succeed][aws-alb-health].

#### Alerts to be removed

We will remove the following alerts that will become superseded:

- [Content Changes unprocessed after 2 hours][unprocessed-content-changes]
- [Messages unprocessed after 2 hours][unprocessed-messages]
- [Incomplete DigestRuns after 2 hours][incomplete-digest-runs]

### Monitoring Email Alert Service

Email Alert Service listens for [Publishing API][] events and communicates
these to Email Alert API. We currently [only have alerts][rabbitmq-alert] that

> occur in-hours if this system fails, so there is no visibility for any

Reviewer: ```suggestion
are monitored in-hours if this system fails, so there is no visibility for any
```

(The alerts still exist out-of-hours.)

alert a support engineer only when the system is exceptionally busy and may be
unhealthy. These alerts will only reach a critical status if they hit a 24 hour
point as this would indicate that we are consistently creating email faster
than the application can send it, which may mean the system is perpetually
backlogged.

#### Web application

We decided that there should be a mechanism that alerts people out-of-hours if
the Email Alert API web application appears broken, noting that currently an
engineer is unlikely to be contacted in the event of the whole Email Alert API
application becoming unavailable - unless travel advice were published. We
noted there wasn't a particularly consistent approach to this across GOV.UK
([with some apps raising an out-of-hours alert when a single instance fails
healthchecks][publishing-api-alert]) and felt it would be good to take a simple
approach to avoid surprising engineers.

We decided the most appropriate place to alert would be when [the application
load balancer has no healthy instances of the application][aws-lb-check]. This
represents that no instances of the web application are healthy and

> [no requests will succeed][aws-alb-health].

Reviewer 2: nit: all the load balancer healthchecks failing doesn't _necessarily_ mean that user requests aren't succeeding, because in that situation the load balancer will behave as if all backends are healthy (because the healthchecks might be returning false negatives so it might as well try — and this type of scenario isn't quite as rare as we might be tempted to imagine).

I think it's still a reasonable surrogate measure (if a little unorthodox) and probably a good tradeoff for now in terms of implementation simplicity. It's also no bad thing to alert in the case where the healthchecks are all broken, even if the service might still be up - since that scenario would mean the alerting is blind to any real outage that might happen until the healthchecks are fixed.

Reviewer 2: The reason I mention this is that it's something worth mentioning when someone comes to write the playbook entry for this alert — i.e. there are two possible causes: either the backends are all unavailable or (less likely but by no means impossible) the healthchecks are broken or timing out and therefore the alert is broken but the service may or may not still be up.

Author: @sengi Thanks for your comments ⭐ 

The behaviour for no healthy instances varies whether you are using an [application load balancer](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/introduction.html) or a [classic load balancer](https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/introduction.html). Key phrasing:

Application Load Balancer: "If a target group contains only unhealthy registered targets, the load balancer nodes route requests across its unhealthy targets."
Classic Load Balancer:  "The load balancer routes requests only to the healthy instances. When the load balancer determines that an instance is unhealthy, it stops routing requests to that instance"

As Email Alert API uses a classic load balancer it will serve 503 responses if all instances of the app are failing the healthcheck - we found this out when the whole app went down and we were wondering where all the stats were (we anticipated high 5xx rate for instance).

Email Alert Frontend is behind an application load balancer so will switch to serving to all instances if all are unhealthy.

---

The tricky bit around this is trying to establish if there is a single point that can be monitored for web application health. If just one box is broke (which admittedly doesn't seem to have much precedent) it doesn't feel necessary to call some up for one, nor is it much fun for a support engineer to get different calls for every box if they all break together. This all seems to tie into smells that are alerting system is a bit outdated as I imagine there are better approaches to all this that the rest of the world have adopted.

Reviewer 2: Oh gosh that's shocking! You're absolutely right though. I hadn't come across that before but it's definitely the case — classic ELB responds `503 Service Unavailable: Back-end server is at capacity` when all its backends are unhealthy.

If I'd known this I would have replaced these with ALBs during the AWS migration 😞

We might want to increase the healthcheck timeout in light of this. It's only 3 seconds at the moment whereas I'm sure the timeout on the app itself is longer than that (the nginx timeout is 15 s, couldn't easily find where the Rails one is set).

Author: Yeah, it was a surprise for me too. To be honest I find it makes quite a lot of sense logically that it's genuinely severe when nothing is healthy - whereas you seem to enter this frustratingly ambiguous world when all healthchecks fail means you're serving all instances (compared to say the oddness of 15 of 16 instances failing healthchecks and you only serving from).

I'm assuming though that "classic" is a synonym for legacy so this behaviour is on it's way out.

There tends to not be a Rails timeout, I think when that is done it is implemented in Rack and I've not seen that we've done that anywhere (we also have plenty of evidence of Rails requests cruising to take > 30s). 15s would sound a better number - I have a feeling that icinga may time out at 10s for healthchecks so could be consistent with that. Mostly the healthchecks should be very fast, but there is a chance they could get stalled if one of their dependent services (a db for instance) has some sort of connection issue.

Reviewer 2: Yeah we actually saw that kind of timeout scenario for real in integration while working on the AWS migration (must have been a different app though because we didn't find out about the surprising blackhole behaviour of the load balancer!) Classic ELB is the first load balancer I've come across which doesn't have some sort of last-resort mode (it's a common enough concept that it has a name). Oh well, the more you know!
